File: sdks/java/src/main/java/io/t402/schemes/svm/SettlementCache.java
package io.t402.schemes.svm;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.time.Instant;
import java.util.HexFormat;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory settlement cache to prevent duplicate concurrent settlement requests.
 * Thread-safe via ConcurrentHashMap. Can be shared across V1/V2 facilitator instances.
 */
public class SettlementCache {

    private static final Duration DEFAULT_TTL = Duration.ofSeconds(60);

    private final Map<String, Instant> entries = new ConcurrentHashMap<>();
    private final Duration ttl;

    public SettlementCache() {
        this(DEFAULT_TTL);
    }

    public SettlementCache(Duration ttl) {
        this.ttl = ttl;
    }

    /**
     * Compute a cache key from transaction bytes.
     */
    public static String transactionKey(byte[] txBytes) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(txBytes);
            return HexFormat.of().formatHex(hash);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    /**
     * Check if a transaction is already being settled.
     * Returns true if duplicate, false if new (and records it).
     */
    public synchronized boolean isDuplicate(String key) {
        pruneExpired();
        if (entries.containsKey(key)) {
            return true;
        }
        entries.put(key, Instant.now());
        return false;
    }

    /**
     * Remove a key from the cache (called after settlement completes).
     */
    public void remove(String key) {
        entries.remove(key);
    }

    /**
     * Current number of entries (for testing).
     */
    public int size() {
        return entries.size();
    }

    private void pruneExpired() {
        Instant cutoff = Instant.now().minus(ttl);
        entries.entrySet().removeIf(entry -> entry.getValue().isBefore(cutoff));
    }
}

File: sdks/go/mechanisms/svm/settlement_cache.go
package svm

import (
	"crypto/sha256"
	"encoding/hex"
	"sync"
	"time"
)

// SettlementCache prevents duplicate concurrent settlement requests for the
// same Solana transaction. It is thread-safe and TTL-based, suitable for
// sharing across V1/V2 facilitator instances.
type SettlementCache struct {
	mu      sync.Mutex
	entries map[string]cacheEntry
	ttl     time.Duration
}

type cacheEntry struct {
	createdAt time.Time
}

// DefaultSettlementTTL is the default time-to-live for cache entries.
const DefaultSettlementTTL = 60 * time.Second

// NewSettlementCache creates a new settlement cache with the given TTL.
// If ttl is 0, DefaultSettlementTTL is used.
func NewSettlementCache(ttl time.Duration) *SettlementCache {
	if ttl == 0 {
		ttl = DefaultSettlementTTL
	}
	c := &SettlementCache{
		entries: make(map[string]cacheEntry),
		ttl:     ttl,
	}
	return c
}

// TransactionKey computes a cache key from raw transaction bytes.
func TransactionKey(txBytes []byte) string {
	h := sha256.Sum256(txBytes)
	return hex.EncodeToString(h[:])
}

// IsDuplicate checks if a transaction is already being settled.
// Returns true if the key is already in the cache (duplicate).
// If not a duplicate, it records the key and returns false.
func (c *SettlementCache) IsDuplicate(key string) bool {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.pruneExpired()

	if _, exists := c.entries[key]; exists {
		return true
	}

	c.entries[key] = cacheEntry{
		createdAt: time.Now(),
	}
	return false
}

// Remove removes a key from the cache (called after settlement completes).
func (c *SettlementCache) Remove(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.entries, key)
}

// Size returns the current number of entries (for testing).
func (c *SettlementCache) Size() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.entries)
}

// pruneExpired removes expired entries. Must be called with mu held.
func (c *SettlementCache) pruneExpired() {
	now := time.Now()
	for key, entry := range c.entries {
		if now.Sub(entry.createdAt) > c.ttl {
			delete(c.entries, key)
		}
	}
}
